### ISSUE TITLE
ruby-vips regressed tests on arm64 and ppc64el

### ISSUE THREAD

**Issue by LocutusOfBorg** [libvips/libvips#1846]
Hello, looks like this commit 8a5dc95fb6848465e609904959f8335684c0a510 from @kleisauke 
Broke the ruby-vips autopkgtests on at least arm64 and s390x
I got the issue after git bisect

and you can see the failure here:
https://autopkgtest.ubuntu.com/packages/r/ruby-vips/groovy/arm64
https://autopkgtest.ubuntu.com/packages/r/ruby-vips/groovy/ppc64el

Basically this is the error 
```
/usr/lib/ruby/vendor_ruby/vips/object.rb:269: warning: Capturing the given block using Proc.new is deprecated; use `&block` instead
......................................F....................................

Failures:

  1) Vips::Image can dilate
     Failure/Error: expect(im.getpoint(11, 12)).to eq([255])

       expected: [255]
            got: [0.0]

       (compared using ==)
```

Can you please have a look?

**Comment by gcsideal** [comment]
Seems this commit is the culprit: https://github.com/libvips/libvips/commit/8a5dc95fb6848465e609904959f8335684c0a510
Code until the previous (and including) commit is fine, so this is the last good state regarding the ruby-vips tests: https://github.com/libvips/libvips/commit/8abcae3abc6ac29db51e559cba18f62562ff872f
Can you double check @LocutusOfBorg ?

**Comment by LocutusOfBorg** [comment]
Yes, because git bisect told me that after 8 runs.

**Comment by LocutusOfBorg** [comment]
```
((8a5dc95f...)|BISECTING) $ git bisect bad
8a5dc95fb6848465e609904959f8335684c0a510 is the first bad commit
commit 8a5dc95fb6848465e609904959f8335684c0a510
Author: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date:   Thu Jun 18 14:29:42 2020 +0200

    Convert the morphology directory to vips8

 libvips/Makefile.am              |   6 +-
 libvips/deprecated/vips7compat.c |  62 +++
 libvips/morphology/Makefile.am   |   1 -
 libvips/morphology/hitmiss.c     | 808 ---------------------------------------
 libvips/morphology/morph.c       | 725 +++++++++++++++++++++++++++++++++--
 libvips/morphology/rank.c        |   2 +-
 po/POTFILES.in                   |   1 -
 7 files changed, 763 insertions(+), 842 deletions(-)
 delete mode 100644 libvips/morphology/hitmiss.c
```

**Comment by LocutusOfBorg** [comment]
I'll check again shortly, in the meanwhile I tried to trigger a build
https://travis-ci.org/github/libvips/ruby-vips/builds/733634889

**Comment by jcupitt** [comment]
Hello @LocutusOfBorg,

Thank you for the detailed report. "Can dilate" failing is probably an error in `hitmiss.c`. It's strange it's not showing up on x86. Could it be a byte order issue?

I'll try on x86 with valgrind.

**Comment by LocutusOfBorg** [comment]
mmm but x86_64 is little endian, as well as ppc64el (el stands just for little endian), and arm64 is little endian too

autopkgtest for ruby-vips/2.0.17-1build1: amd64: Pass, arm64: Regression ♻ , armhf: Pass, i386: Ignored failure, ppc64el: Regression ♻ , s390x: Not a regression

This is what I see, and I can't explain. armhf is 32bit arm (hard floating point), but same endianess as arm64, and its good.

s390x is the only big endian platform we have in Ubuntu, and this test fails too (as well as another one, so that one is probably an endianess issue) https://objectstorage.prodstack4-5.canonical.com/v1/AUTH_77e2ada1e7a84929a74ba3b87153c0ac/autopkgtest-groovy/groovy/s390x/r/ruby-vips/20201006_023326_d446f@/log.gz

**Comment by jcupitt** [comment]
I've reproduced the error on x86! This prog:

```
#!/bin/bash
  
vips black x.v 16 16
vips draw_rect x.v 255 10 12 1 1
cat >mask.v <<EOF
3 3
128 255 128
255 255 255
128 255 128
EOF
valgrind vips morph x.v x2.v mask.v dilate --vips-novector
echo -n "should be 255: "
vips getpoint x2.v 10 12
echo -n "should be 255: "
vips getpoint x2.v 11 12
echo -n "should be 0: "
vips getpoint x2.v 12 12
```

Looks like an error in the C path. It runs on the vector path on x86 so the error didn't show up. I'll investigate.

**Comment by LocutusOfBorg** [comment]
nice! also the endianess test error is clearly shown in travis
https://travis-ci.org/github/kleisauke/ruby-vips/jobs/733637121

```
  1) Vips::Image can load an image from a file

     Failure/Error: expect(x.avg).to eq(128)

       expected: 128

            got: 9.388699710976274e-44

       (compared using ==)

     # ./spec/image_spec.rb:27:in `block (2 levels) in <top (required)>'
```

**Comment by kleisauke** [comment]
Commit https://github.com/kleisauke/libvips/commit/143c815a0a85cb187bc6f690dab0555fd617063e seems to fix it for me. Sorry for causing this regression.

**Comment by LocutusOfBorg** [comment]
thanks for the quick fix! I wasn't expecting it to be that fast!
I thought the Debian maintainer was going to report it upstream, but probably he missed my bug report, I'm happy to have done it myself :D

**Comment by gcsideal** [comment]
> thanks for the quick fix! I wasn't expecting it to be that fast!
> I thought the Debian maintainer was going to report it upstream, but probably he missed my bug report, I'm happy to have done it myself :D

Nope, I've seen it back then (the Debian bug report) and was in contact with upstream in private about this. Should have been visible, sorry for that.

**Comment by LocutusOfBorg** [comment]
> > thanks for the quick fix! I wasn't expecting it to be that fast!
> > I thought the Debian maintainer was going to report it upstream, but probably he missed my bug report, I'm happy to have done it myself :D
> 
> Nope, I've seen it back then (the Debian bug report) and was in contact with upstream in private about this. Should have been visible, sorry for that.

no problem at all :) I'm happy to have speeded it up a little bit :)

**Comment by LocutusOfBorg** [comment]
btw I already uploaded in Ubuntu
https://launchpad.net/ubuntu/+source/vips/8.10.1-2ubuntu1
in ~1h we will have results here https://autopkgtest.ubuntu.com/packages/r/ruby-vips/groovy/ppc64el

**Comment by jcupitt** [comment]
Yes, fixes it for me too.

It's a clear error once you see it (a `++` had been dropped in the recent refactoring), so I've committed to the 8.10 branch and master. I didn't use your patch exactly Kleis, I wanted to keep all changes to `t` close together for clarity.

Nice find!

**Comment by gcsideal** [comment]
> btw I already uploaded in Ubuntu
> https://launchpad.net/ubuntu/+source/vips/8.10.1-2ubuntu1
> in ~1h we will have results here https://autopkgtest.ubuntu.com/packages/r/ruby-vips/groovy/ppc64el

Shame on me, but I believe the opposite path. Test first and upload only if it works. Just got the results, it's a correct fix. Still, going to use the one from upstream.

**Comment by LocutusOfBorg** [comment]
> > btw I already uploaded in Ubuntu
> > https://launchpad.net/ubuntu/+source/vips/8.10.1-2ubuntu1
> > in ~1h we will have results here https://autopkgtest.ubuntu.com/packages/r/ruby-vips/groovy/ppc64el
> 
> Shame on me, but I believe the opposite path. Test first and upload only if it works. Just got the results, it's a correct fix. Still, going to use the one from upstream.

I can't test every architecture where it fails :) my testing has been performed on an arm64-based qemu chroot, as shown on the whole bug report. but since the failure is the same for ppc64el, I guess also the fix is!

**Comment by LocutusOfBorg** [comment]
and it worked also on Ubuntu infrastructure :)

**Comment by jcupitt** [comment]
Great! I'll leave it a few days in case anything else comes up, and then push out 8.10.2.

Thank you everyone for testing!

**Comment by jcupitt** [comment]
(and reporting, of course)

**Comment by LocutusOfBorg** [comment]
thanks you all!

**Comment by jcupitt** [comment]
Should be fixed in 8.10.2 https://github.com/libvips/libvips/releases/tag/v8.10.2
